cmd/test-server/zerotrace.go

```diff
@@ -0,0 +1,296 @@
+	clientPort, _ := strconv.Atoi(clPort)
+
+	return &zeroTrace{
+		Iface:      iface,
+		Conn:       conn,
+		ClientIP:   clientIP,
+		ClientPort: clientPort,
+	}
+}
+
+// sendTTLIncrementingProbes sends probes of incrementing TTL, await response from channel that identifies the hop which sent the ICMP response and stops sending any more probes if connection errors out
+func (z *zeroTrace) sendTTLIncrementingProbes(recvdHopData chan HopRTT) (map[int]HopRTT, error) {
+	traceroute := make(map[int]HopRTT)
+	for ttlValue := beginTTLValue; ttlValue <= maxTTLHops; ttlValue++ {
+		if err := z.sendTracePacket(ttlValue); err != nil {
+			ErrLogger.Println("Send Trace Packet Error: ", err)
+			return traceroute, err
+		}
+		z.CurrTTLIndicator = ttlValue
+		ticker := time.NewTicker(tracerouteHopTimeout)
+		defer ticker.Stop()
+		select {
+		case hopData := <-recvdHopData:
+			traceroute[ttlValue] = hopData
+		case <-ticker.C:
+			ErrLogger.Println("Traceroute Hop Timeout at Hop ", ttlValue, ". Moving on to the next hop.")
+			var empty net.IP
+			traceroute[ttlValue] = HopRTT{empty, 0}
+			continue
+		}
+		if traceroute[ttlValue].IP.String() == z.ClientIP {
+			break
+		}
+	}
+	return traceroute, nil
+}
+
+// Run reaches the underlying connection and sets up necessary pcap handles
+// and implements the 0trace method of sending TTL-limited probes on an existing TCP connection
+func (z *zeroTrace) Run() (map[int]HopRTT, error) {
+	var err error
+	if z.PcapHdl, err = z.setupPcapAndFilter(); err != nil {
+		return nil, err
+	}
+
+	recvdHopChan := make(chan HopRTT)
+	quit := make(chan bool)
+	// Fire go routine to start listening for packets on the handler before sending TTL limited probes
+	go z.recvPackets(z.PcapHdl, recvdHopChan, quit)
+
+	traceroute, err := z.sendTTLIncrementingProbes(recvdHopChan)
+	quit <- true
+	return traceroute, err
+}
+
+// setupPcap sets up the pcap handle on the required interface, and applies the filter and returns it
+func (z *zeroTrace) setupPcapAndFilter() (*pcap.Handle, error) {
+	pcapHdl, err := pcap.OpenLive(z.Iface, snaplen, promisc, time.Second)
+	if err != nil {
+		return nil, err
+	}
+	if err = pcapHdl.SetBPFFilter(fmt.Sprintf("(tcp and port %d and host %s) or icmp", z.ClientPort, z.ClientIP)); err != nil {
+		return nil, err
+	}
+	return pcapHdl, nil
+}
+
+// recvPackets listens on the provided pcap handler for packets sent, processes TCP and ICMP packets differently, and aborts if signalled
+func (z *zeroTrace) recvPackets(pcapHdl *pcap.Handle, hops chan HopRTT, quit chan bool) {
+	tempConn := z.Conn.(*tls.Conn)
+	tcpConn := tempConn.NetConn()
+
+	localSrcAddr := tcpConn.LocalAddr().String()
+	serverIP, _, _ := net.SplitHostPort(localSrcAddr)
+
+	z.SentPktsIPId = make(map[int][]SentPacketData)
+	packetStream := gopacket.NewPacketSource(pcapHdl, pcapHdl.LinkType())
+	var counter int
+
+	for packet := range packetStream.Packets() {
+		select {
+		case <-quit:
+			return
```

Comment:
I've found an issue with this:
In this file L:113, we feed "true" to the quit channel once the `sendTTLIncrementingProbes` function returns. But since I am working with the packet stream here, I need to wait until another packet is seen on this handle before being able to quit. Moving the select statement out also does not work since the execution will be stuck in the `for packet := range packetStream.Packets()` loop. 

Trying to fix this now

Reply:
A fix has been in a89c1d2

Reply:
That's the same fix that I had in mind when reviewing the code :+1: